src/workerd/api/encoding.c++

```diff
@@ -460,29 +461,204 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
+        pendingSurrogate = false;
+      } else {
+        position += encodeUtf8CodeUnit(0xFFFD, out + position);
+        if (!isLeadSurrogate(c)) {
+          position += encodeUtf8CodeUnit(c, out + position);
+          pendingSurrogate = false;
+        }
+      }
+    } else if (isLeadSurrogate(c)) {
+      pendingSurrogate = true;
+    } else {
+      if (isTrailSurrogate(c)) {
+        position += encodeUtf8CodeUnit(0xFFFD, out + position);
+      } else {
+        position += encodeUtf8CodeUnit(c, out + position);
+      }
+    }
+  }
+
+  if (pendingSurrogate) {
+    encodeUtf8CodeUnit(0xFFFD, out + position);
+  }
+}
+
 }  // namespace
 
-jsg::BufferSource TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
-  auto str = input.orDefault(js.str());
-  auto view = JSG_REQUIRE_NONNULL(jsg::BufferSource::tryAlloc(js, str.utf8Length(js)), RangeError,
-      "Cannot allocate space for TextEncoder.encode");
-  [[maybe_unused]] auto result = encodeIntoImpl(js, str, view);
-  KJ_DASSERT(result.written == view.size());
-  return kj::mv(view);
+jsg::Ref<TextEncoder> TextEncoder::constructor(jsg::Lock& js) {
+  return js.alloc<TextEncoder>();
+}
+
+jsg::JsUint8Array TextEncoder::encode(jsg::Lock& js, jsg::Optional<jsg::JsString> input) {
+  jsg::JsString str = input.orDefault(js.str());
+  std::shared_ptr<v8::BackingStore> backingStore;
+  size_t utf8_length = 0;
+
+  // Fast path: check if string is one-byte before creating ValueView
+  if (str.isOneByte(js)) {
+    auto length = str.length(js);
+    // Allocate buffer for Latin-1. Use v8::ArrayBuffer::NewBackingStore to avoid creating
+    // JS objects during conversion.
+    backingStore = v8::ArrayBuffer::NewBackingStore(
+        js.v8Isolate, length, v8::BackingStoreInitializationMode::kUninitialized);
+    auto backingData = reinterpret_cast<kj::byte*>(backingStore->Data());
+
+    str.writeOneByte(js, kj::ArrayPtr<kj::byte>(backingData, length),
+        jsg::JsString::WriteFlags::REPLACE_INVALID_UTF8);
+
+    utf8_length =
+        simdutf::utf8_length_from_latin1(reinterpret_cast<const char*>(backingData), length);
+
+    if (utf8_length == length) {
+      // ASCII fast path: no conversion needed, Latin-1 is same as UTF-8 for ASCII
+      auto array = v8::Uint8Array::New(v8::ArrayBuffer::New(js.v8Isolate, backingStore), 0, length);
+      return jsg::JsUint8Array(array);
+    }
+
+    // Need to convert Latin-1 to UTF-8
+    std::shared_ptr<v8::BackingStore> backingStore2 = v8::ArrayBuffer::NewBackingStore(
+        js.v8Isolate, utf8_length, v8::BackingStoreInitializationMode::kUninitialized);
+    auto written = simdutf::convert_latin1_to_utf8(reinterpret_cast<const char*>(backingData),
+        length, reinterpret_cast<char*>(backingStore2->Data()));
+    KJ_DASSERT(utf8_length == written);
+    auto array =
+        v8::Uint8Array::New(v8::ArrayBuffer::New(js.v8Isolate, backingStore2), 0, utf8_length);
+    return jsg::JsUint8Array(array);
```

Comment:
This is still creating the intermediate on-heap allocation that can trigger GC. Let's have the intermediate allocation (the one that `writeOneByte` writes into) be an off-heap (`kj::SmallArray<kj::byte, 4096> buffer(length)`) allocation. In both cases we end up copying into the on-heap destination buffer but we avoid the wasted additional on-heap allocation in the second case.

Reply:
It's not about the memcpy... on line 601 you allocate a v8::BackingStore that receives the results of the WriteOneByte, then here in this branch you create a second one that receives the results of the latin1-to-utf8 encoding. What I'm saying here is that you should only create one v8::BackingStore that represents the final result. The WriteOneByte should write into an off-v8-heap allocation (e.g. a kj::Array). This way you greatly lessen the risk of triggering a GC in the middle.

Reply:
The advantage of using BackingStore is that if it's ASCII you are already done and don't have to move it into the sandbox.

Reply:
Yes, I get why it's being used here but that's missing the point: the issue is that `v8::BackingStore` can trigger GC, even if it is not technically itself an on-heap allocation. It might be rare enough to not worry about, which would be great to demonstrate, but there's an alternative approach we can take here that we *know* won't trigger GC and from the flamegraphs we're looking at the GC costs appear to far outweight the copy costs

The issue we're having with encode in certain benchmarks is additional memory pressure caused by flattening and the additional GC-triggering allocations.
<img width="3964" height="712" alt="image" src="https://github.com/user-attachments/assets/6bffff1d-7ccf-44f8-9a2f-c1c11faf8889" />